sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,835 @@
+            options = self._ack_map.pop(ack_id)
+            if options.error_detail is not None:
+                raise SendMessageError(
+                    message="Failed to send message.", ack_id=options.ack_id, error_detail=options.error_detail
+                )
+
+    def _get_or_add_group(self, name: str) -> WebPubSubGroup:
+        if name not in self._group_map:
+            self._group_map[name] = WebPubSubGroup(name=name)
+        return self._group_map[name]
+
+    @distributed_trace
+    def join_group(self, group_name: str, **kwargs: Any) -> None:
+        """Join the client to group.
+
+        :param group_name: The group name. Required.
+        :type group_name: str.
+        :keyword int ack_id: The optional ackId. If not specified, client will generate one.
+        """
+
+        def join_group_attempt():
+            group = self._get_or_add_group(group_name)
+            self._join_group_core(group_name, **kwargs)
+            group.is_joined = True
+
+        self._retry(join_group_attempt)
+
+    def _join_group_core(self, group_name: str, **kwargs: Any) -> None:
+        ack_id = kwargs.pop("ack_id", None)
+        self._send_message_with_ack_id(
+            message_provider=lambda id: JoinGroupMessage(group=group_name, ack_id=id),
+            ack_id=ack_id,
+            **kwargs,
+        )
+
+    @distributed_trace
+    def leave_group(self, group_name: str, **kwargs: Any) -> None:
+        """Leave the client from group
+        :param group_name: The group name. Required.
+        :type group_name: str.
+        :keyword int ack_id: The optional ackId. If not specified, client will generate one.
+        """
+
+        def leave_group_attempt():
+            group = self._get_or_add_group(group_name)
+            ack_id = kwargs.pop("ack_id", None)
+            self._send_message_with_ack_id(
+                message_provider=lambda id: LeaveGroupMessage(group=group_name, ack_id=id),
+                ack_id=ack_id,
+                **kwargs,
+            )
+            group.is_joined = False
+
+        self._retry(leave_group_attempt)
+
+    @distributed_trace
+    def send_event(
+        self,
+        event_name: str,
+        content: Any,
+        data_type: Union[WebPubSubDataType, str],
+        **kwargs: Any,
+    ) -> None:
+        """Send custom event to server
+        :param event_name: The event name. Required.
+        :type event_name: str.
+        :param content: The data content. Required.
+        :type content: Any.
+        :param data_type: The data type. Required.
+        :type data_type: Union[WebPubSubDataType, str].
+        :keyword int ack_id: The optional ackId. If not specified, client will generate one.
+        :keyword bool fire_and_forget: If true, the message won't contains ackId and no AckMessage
+         will be returned from the service.
+        """
+
+        def send_event_attempt():
+            fire_and_forget = kwargs.pop("fire_and_forget", False)
```

Comment:
right name for a kwarg? I know it's not documented but it is technically public

Reply:
I declar it in doc string: https://github.com/Azure/azure-sdk-for-python/blob/f392fb5871d95546c16ea6ac399e93d786036582/sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py#L286-L287

Reply:
I feel like there might be a cleaner name for this... like just `ack=True` - `fire_and_forget` is very long for a pretty simply concept. I wonder what other libraries are doing here.

Reply:
I think `ack` is fine.